### PR TITLE
Feature scenario celery

### PR DESCRIPTION
**Pull request by sadrakhamoshi**

### CODE REVIEW

File: mhq_web/settings.py [hunk 1]
```diff
@@ -170,6 +170,14 @@
 EMAIL_HOST_PASSWORD = secret_key_gmail
 EMAIL_USE_TLS = True
 
+CELERY_BROKER_URL = 'redis://redis:6379'
+CELERY_RESULT_BACKEND = 'redis://redis:6379'
+CELERY_ACCEPT_CONTENT = ['application/json']
+CELERY_TASK_SERIALIZER = 'json'
+CELERY_RESULT_SERIALIZER = 'json'
+CELERY_TIMEZONE = 'Asia/Tehran'
```
Comment (AmirH-KHALI) [inline mhq_web/settings.py:178]: khob ina ke hamun bala hast🤔

File: condition/tasks.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from celery import shared_task
+
+
+@shared_task
+def hello():
+    print("Hello there!fasfsadf;adsjfds ")
```
Comment (AmirH-KHALI) [inline condition/tasks.py:6]: khob in ke pak she tabiatan

File: module/models.py [hunk 3]
```diff
@@ -9,3 +10,10 @@ class Module(models.Model):
 
     class Meta:
         db_table = 'module'
+
+    def get_neighbours(self):
+        near = Edge.objects.all().filter(source=self)
+        return [n.dist for n in near]
+
+    def run_module(self):
+        print('Hello')
```
Comment (AmirH-KHALI) [inline module/models.py:19]: inam pak

File: request/managers.py [hunk 4]
```diff
@@ -55,9 +57,19 @@ def execute(self):
...
+        req = RequestHistory.objects.create(user=self.user,
+                                            name=self.request.name,
+                                            http_method=self.request.http_method,
+                                            url=self.request.url,
+                                            body=self.request.body,
+                                            headers=get_key_value_dict(self.request.get_headers()),
+                                            params=get_key_value_dict(self.request.get_params()),
+                                            response=response)
+        req.save()
+        return response, req.id
```
Comment (AmirH-KHALI) [inline request/managers.py:75]: ino harf bezanim ke chera inja bashe history khordan e🤔

File: scenario/execute.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+from edge.models import Edge
+from request.managers import RequestExecution
+from .signals import scenario_run_finished
+
+
+class Execution:
```
Comment (AmirH-KHALI) [inline scenario/execute.py:6]: esm e class be nazaram ScenarioExecution bashe

File: scenario/execute.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            try:
+                response, req_id = RequestExecution(request=request, user=self.user).execute()
+            except Exception as e:
+                return e, module
+            self.response_list.append((module.id, response))
+            responses.append(response)
+            self.path.append(req_id)
+            edges = Edge.objects.all().filter(source=module)
+            for e in edges:
+                if self.check_statement(response, e.dist):
```
Comment (AmirH-KHALI) [inline scenario/execute.py:33]: tu check statement edge migire vali inja e.dist dadi

File: scenario/permissions.py [hunk 7]
```diff
@@ -21,3 +23,19 @@ def has_object_permission(self, request, view, obj):
...
+            if collection.type == Collection.PRIVATE:
+                get_object_or_404(UserCollection, user=request.user, collection=collection)
+            return True
+
+        if request.method == 'DELETE':
+            collection = get_object_or_404(ScenarioHistory, id=view.kwargs.get('pk')).scenario.collection
+            user_collection = get_object_or_404(UserCollection, user=request.user, collection=collection)
+            if user_collection.role == UserCollection.VISITOR:
+                return False
+        return True
```
Comment (AmirH-KHALI) [inline scenario/permissions.py:41]: inam hal nadashtam bekhunam, vali khob get_object_or_404 nabas tu permission bashe dige🤔

File: scenario/serializers.py [hunk 8]
```diff
@@ -105,3 +108,25 @@ class ModuleScenarioSerializer(serializers.ModelSerializer):
...
         fields = ['id', 'request', 'x_position', 'y_position']
+
+
+class ScenarioHistorySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ScenarioHistory
+        fields = '__all__'
+
+    def validate(self, attrs):
+        print("Hello")
```
Comment (AmirH-KHALI) [inline scenario/serializers.py:119]: pak

File: scenario/serializers.py [hunk 9]
```diff
@@ -105,3 +108,25 @@ class ModuleScenarioSerializer(serializers.ModelSerializer):
...
+        if len(attrs['order']) == 0 and ('schedule' not in attrs):
```
Comment (AmirH-KHALI) [inline scenario/serializers.py:120]: attrs nemdunam chie tozih mikhad

File: scenario/serializers.py [hunk 10]
```diff
@@ -105,3 +108,25 @@ class ModuleScenarioSerializer(serializers.ModelSerializer):
...
+            raise serializers.ValidationError('you should pass order or schedule')
+        if len(attrs['order']) != 0:
+            for o in attrs['order']:
+                get_object_or_404(RequestHistory, id=o)
```
Comment (AmirH-KHALI) [inline scenario/serializers.py:124]: inja ham nabas get_object_or_404 bashe, in faqat tu viewwww 😂😂😂😂

File: scenario/signals.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    scenario = kwargs.get('scenario')
+    order = kwargs.get('order')
+    # exe_time = kwargs.get('execution_time')
+    # schedule = kwargs.get('schedule')
+
+    ScenarioHistory.objects.create(name=name,
+                                   user=user,
+                                   scenario=scenario,
+                                   order=order,
+                                   collection=scenario.collection).save()
```
Comment (AmirH-KHALI) [inline scenario/signals.py:23]: signal kolan nemdunam bas tozih bedi

File: scenario/tasks.py [hunk 12]
```diff
@@ -1,8 +1,17 @@
...
-    scenario = Scenario.objects.get(id=scenario_id)
-    print('Hi')
+def execute_scenario():
+    res = requests.post('https://60c1f3514f7e880017dc0e65.mockapi.io/scenario')
+    print(res)
+    # exe = Execution(scenario=scenario)
+    # response = exe.execute()
+    # if isinstance(response, Exception):
+    #     return str(response)
+    # return response
```
Comment (AmirH-KHALI) [inline scenario/tasks.py:17]: comment ha pak

File: scenario/tasks.py [hunk 13]
```diff
@@ -1,8 +1,17 @@
...
+import requests
 
 
 @app.task
-def execute(scenario_id):
-    scenario = Scenario.objects.get(id=scenario_id)
-    print('Hi')
+def execute_scenario():
+    res = requests.post('https://60c1f3514f7e880017dc0e65.mockapi.io/scenario')
+    print(res)
```
Comment (AmirH-KHALI) [inline scenario/tasks.py:12]: ow in asan be scenario execution vasl nashode

File: scenario/urls.py [hunk 14]
```diff
@@ -9,4 +9,7 @@
     path('<int:pk>/set_starter_module/<int:module_id>/', viewsets.scenario_starter_module),
     path('<int:pk>/schedule/', viewsets.scenario_schedule),
     path('all_modules/<int:scenario_id>', viewsets.all_modules_scenario),
+    path('execute/<int:scenario_id>', viewsets.execute),
```
Comment (AmirH-KHALI) [inline scenario/urls.py:12]: ye kari kon con url injuri hast ke "{IP_ADDRESS}/api/scenario/", bad e in scenario e id ish biad bad bezanim execute masalan behtare be nazaram, hala mishe bahs kard

File: scenario/urls.py [hunk 15]
```diff
@@ -9,4 +9,7 @@
...
+    path('scenario_history/<int:pk>/', viewsets.scenario_history_delete),
```
Comment (AmirH-KHALI) [inline scenario/urls.py:13]: inam mese balaii, va chera ye jahaii pk, ye jahaii scenario_id, ye chiz bezar hamaro dige lamasab😂

File: scenario/views.py [hunk 16]
```diff
@@ -58,6 +63,30 @@ def set_starter_module(self, request, pk, module_id):
...
 
+    def execute(self, request, *args, **kwargs):
+        scenario_id = kwargs.get('scenario_id')
+        scenario = get_object_or_404(Scenario, id=scenario_id)
+        exe = Execution(scenario=scenario, user=request.user)
+        response = exe.execute()
+        if len(response) == 2 and isinstance(response[0], Exception):
+            return Response({'msg': str(response[0]),
+                             'request': str(response[1])}, status=status.HTTP_400_BAD_REQUEST)
+        return Response({'msg': exe.response_list, 'path': exe.path}, status=status.HTTP_200_OK)
```
Comment (AmirH-KHALI) [inline scenario/views.py:74]: ina ro ham fek konam ezafi bara test zadi🤔, chon age celery biad dige chizi qarar nist khoruji bedim🤔

File: condition/tasks.py (hunk 2 above)
Comment (sadrakhamoshi) [reply condition/tasks.py:6]: Re in byad pak she

File: mhq_web/settings.py (hunk 1 above)
Comment (sadrakhamoshi) [reply mhq_web/settings.py:178]: re pak shod

File: module/models.py (hunk 3 above)
Comment (sadrakhamoshi) [reply module/models.py:19]: hal

File: request/managers.py (hunk 4 above)
Comment (sadrakhamoshi) [reply request/managers.py:75]: harf miznim saresh

File: scenario/execute.py (hunk 5 above)
Comment (sadrakhamoshi) [reply scenario/execute.py:6]: movafegham

File: scenario/execute.py (hunk 6 above)
Comment (sadrakhamoshi) [reply scenario/execute.py:33]: fk konm commit nashode inm okaye

File: scenario/permissions.py (hunk 7 above)
Comment (sadrakhamoshi) [reply scenario/permissions.py:41]: re b nazrm permission ha ro hmasho ye chiz bezarim chon hamshon ye kar byd bokonn taghribn vase hame chiz hye scenario

File: scenario/serializers.py (hunk 10 above)
Comment (sadrakhamoshi) [reply scenario/serializers.py:124]: hallllllllllllllll

File: scenario/signals.py (hunk 11 above)
Comment (sadrakhamoshi) [reply scenario/signals.py:23]: hale

File: scenario/tasks.py (hunk 13 above)
Comment (sadrakhamoshi) [reply scenario/tasks.py:12]: re dg fln be celery vasl nashode fuck dockerrrr

File: scenario/tasks.py (hunk 12 above)
Comment (sadrakhamoshi) [reply scenario/tasks.py:17]: to commit badi byd vasl beshe b celery

File: scenario/urls.py (hunk 15 above)
Comment (sadrakhamoshi) [reply scenario/urls.py:13]: bahss

**Pull request merged**